pkg/controller/podprobemarker/node_pod_probe.go

```diff
@@ -0,0 +1,311 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podprobemarker
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	nodeutil "k8s.io/kubernetes/pkg/controller/util/node"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	nodeImageCreationDelayAfterNodeReady = time.Second * 30
+)
+
+func (r *ReconcilePodProbeMarker) syncNodePodProbe(name string) error {
+	// Fetch the NodePodProbe instance
+	npp := &appsv1alpha1.NodePodProbe{}
+	err := r.Get(context.TODO(), client.ObjectKey{Name: name}, npp)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			npp.Name = name
+			return r.Create(context.TODO(), npp)
+		}
+		// Error reading the object - requeue the request.
+		return err
+	}
+	// Fetch Node
+	node := &corev1.Node{}
+	err = r.Get(context.TODO(), client.ObjectKey{Name: name}, node)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	} else if errors.IsNotFound(err) || !node.DeletionTimestamp.IsZero() {
+		return r.Delete(context.TODO(), npp)
+	}
+	// If Pod is deleted, then remove podProbe from NodePodProbe.Spec
+	matchedPods, ppms, err := r.syncPodFromNodePodProbe(npp)
+	if err != nil {
+		return err
+	}
+	for _, status := range npp.Status.PodProbeStatuses {
```

Comment:
npp.Status can be an empty slice, how to init npp status and how to add probestatus to existing status if new pods are added

Reply:
npp.Status is daemon to update